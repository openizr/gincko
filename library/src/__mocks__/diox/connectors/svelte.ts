/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { readable } from 'svelte/store/index';

/**
 * diox/connectors/svelte mock.
 */
const connect = vi.fn(() => {
  const state = {
    steps: [{ id: 'start' }, { id: 'end' }],
    loading: process.env.LOADING === 'true',
  };
  return vi.fn(() => readable(state, (set) => {
    set(state);
  }));
});

export default connect;
