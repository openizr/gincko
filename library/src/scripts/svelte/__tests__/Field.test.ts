/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @jest-environment jsdom
 */

import Field from 'scripts/svelte/Field.svelte';
import { render, waitFor } from '@testing-library/svelte';

jest.mock('biuty/svelte');
jest.mock('scripts/core/Engine');
jest.mock('scripts/svelte/Message.svelte');
jest.mock('scripts/svelte/NestedFields.svelte');

const flushPromise = jest.fn(() => new Promise<void>((resolve) => { setTimeout(resolve, 50); }));

describe('svelte/Field', () => {
  const isActive = true;
  const customComponents = {};
  const userInputs = { test: 'ok' };
  const variables = { var: 'value' };
  const i18n = jest.fn(() => 'TRANSLATED LABEL');
  const onUserAction = jest.fn() as unknown as OnUserAction;

  beforeEach(() => {
    jest.clearAllMocks();
    delete process.env.IS_DATE;
  });

  test('renders correctly - Unknown component', async () => {
    const { container } = render(Field, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          status: 'success',
          component: 'Unknown',
          componentProps: {},
        },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Message with label and helper', async () => {
    const { container } = render(Field, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          label: 'Test',
          status: 'success',
          component: 'Message',
          componentProps: { helper: 'Helper' },
        },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Message with label and message', async () => {
    const { container } = render(Field, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          label: 'Test',
          message: 'Success',
          status: 'success',
          component: 'Message',
          componentProps: { helper: 'Helper' },
        },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Button', async () => {
    const { container } = render(Field, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          status: 'success',
          component: 'Button',
          componentProps: {},
        },
      },
    });
    await waitFor(flushPromise);
    expect(container.firstChild).toMatchSnapshot();
    expect(onUserAction).toHaveBeenCalledTimes(1);
    expect(onUserAction).toHaveBeenCalledWith('input', 'path.0.to.field', true);
  });

  test('renders correctly - Link', async () => {
    const { container } = render(Field, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          status: 'success',
          component: 'Link',
          componentProps: {},
        },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Options', async () => {
    const { container } = render(Field, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          status: 'success',
          component: 'Options',
          componentProps: {
            onFocus: (): null => null,
            options: [{ label: 'test', value: 'option1' }, { value: 'option2' }],
          },
        },
      },
    });
    await waitFor(flushPromise);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - DatePicker', async () => {
    process.env.IS_DATE = 'true';
    const { container, rerender } = render(Field, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          status: 'success',
          component: 'DatePicker',
          componentProps: {},
          value: null,
        },
      },
    });
    await waitFor(flushPromise);
    expect(container.firstChild).toMatchSnapshot();
    rerender({
      isActive,
      i18n,
      path: 'path.0.to.field',
      variables,
      userInputs,
      onUserAction,
      customComponents,
      field: {
        id: 'field',
        status: 'success',
        component: 'DatePicker',
        value: new Date(1657192371401),
        componentProps: { placeholder: 'placeholder', debounceTimeout: 10 },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Textfield', async () => {
    const { container, rerender } = render(Field, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          value: null,
          status: 'success',
          component: 'Textfield',
          componentProps: {},
        },
      },
    });
    await waitFor(flushPromise);
    expect(container.firstChild).toMatchSnapshot();
    rerender({
      isActive,
      i18n,
      path: 'path.0.to.field',
      variables,
      userInputs,
      onUserAction,
      customComponents,
      field: {
        id: 'field',
        value: 'test',
        status: 'success',
        component: 'Textfield',
        componentProps: { placeholder: 'placeholder', debounceTimeout: 10 },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Textarea', async () => {
    const { container, rerender } = render(Field, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          value: null,
          status: 'success',
          component: 'Textarea',
          componentProps: {},
        },
      },
    });
    await waitFor(flushPromise);
    expect(container.firstChild).toMatchSnapshot();
    rerender({
      isActive,
      i18n,
      path: 'path.0.to.field',
      variables,
      userInputs,
      onUserAction,
      customComponents,
      field: {
        id: 'field',
        value: 'test',
        status: 'success',
        component: 'Textarea',
        componentProps: { placeholder: 'placeholder', debounceTimeout: 10 },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - FilePicker', async () => {
    const { container, rerender } = render(Field, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          status: 'success',
          component: 'FilePicker',
          componentProps: {},
        },
      },
    });
    await waitFor(flushPromise);
    expect(container.firstChild).toMatchSnapshot();
    rerender({
      isActive,
      i18n,
      path: 'path.0.to.field',
      variables,
      userInputs,
      onUserAction,
      customComponents,
      field: {
        id: 'field',
        status: 'success',
        component: 'FilePicker',
        componentProps: { placeholder: 'placeholder' },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Array', async () => {
    const { container, rerender } = render(Field, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          status: 'success',
          component: 'Array',
          componentProps: {},
          fields: [],
        },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
    rerender({
      isActive: false,
      i18n,
      path: 'path.0.to.field',
      variables,
      userInputs,
      onUserAction,
      customComponents,
      field: {
        id: 'field',
        status: 'success',
        component: 'Array',
        componentProps: {
          addButtonProps: { label: 'add' },
          removeButtonProps: { label: 'remove' },
        },
        fields: [],
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - DynamicObject', async () => {
    const { container, rerender } = render(Field, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          status: 'success',
          component: 'DynamicObject',
          componentProps: {},
          fields: [],
        },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
    rerender({
      isActive,
      i18n,
      path: 'path.0.to.field',
      variables,
      userInputs,
      onUserAction,
      customComponents,
      field: {
        id: 'field',
        status: 'success',
        component: 'DynamicObject',
        componentProps: {
          addTextfieldProps: { label: 'pattern', placeholder: 'placeholder' },
        },
        fields: [],
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
