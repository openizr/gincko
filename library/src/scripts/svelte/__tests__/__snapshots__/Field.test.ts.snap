// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`svelte/Field renders correctly - Array 1`] = `
<div>
  <div
    id="field"
  >
    {"id":"path__0__to__field","path":"path.0.to.field","fields":[],"minItems":0,"modifiers":"success ","isActive":true,"maxItems":null,"variables":{"var":"value"},"userInputs":{"test":"ok"},"label":null,"helper":null,"allowedPatterns":[],"addButtonProps":{"label":null},"removeButtonProps":{"label":null},"addTextfieldProps":{"label":null,"placeholder":null},"type":"array","value":null,"customComponents":{}}
  </div>
  
  
</div>
`;

exports[`svelte/Field renders correctly - Array 2`] = `
<div>
  <div
    id="field"
  >
    {"id":"path__0__to__field","path":"path.0.to.field","fields":[],"minItems":0,"modifiers":"success ","isActive":false,"maxItems":null,"variables":{"var":"value"},"userInputs":{"test":"ok"},"label":null,"helper":null,"allowedPatterns":[],"addButtonProps":{"label":"TRANSLATED LABEL"},"removeButtonProps":{"label":"TRANSLATED LABEL"},"addTextfieldProps":{"label":null,"placeholder":null},"type":"array","value":null,"customComponents":{}}
  </div>
  
  
</div>
`;

exports[`svelte/Field renders correctly - Button 1`] = `
<div>
  <div>
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    null
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    path__0__to__field
     
    undefined
     
    success 
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
  </div>
  
  
</div>
`;

exports[`svelte/Field renders correctly - DatePicker 1`] = `
<div>
  <div>
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    path.0.to.field
     
    null
     
    null
     
    null
     
    null
     
    100
     
    path__0__to__field
     
    false
     
    success 
     
    function
     
    10
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
  </div>
  
  
</div>
`;

exports[`svelte/Field renders correctly - DatePicker 2`] = `
<div>
  <div>
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    path.0.to.field
     
    null
     
    2022/07/07
     
    TRANSLATED LABEL
     
    null
     
    10
     
    path__0__to__field
     
    false
     
    success 
     
    function
     
    10
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
  </div>
  
  
</div>
`;

exports[`svelte/Field renders correctly - DynamicObject 1`] = `
<div>
  <div
    id="field"
  >
    {"id":"path__0__to__field","path":"path.0.to.field","fields":[],"minItems":0,"modifiers":"success ","isActive":true,"maxItems":null,"variables":{"var":"value"},"userInputs":{"test":"ok"},"label":null,"helper":null,"allowedPatterns":[],"addButtonProps":{"label":null},"removeButtonProps":{"label":null},"addTextfieldProps":{"label":null,"placeholder":null},"type":"dynamicObject","value":null,"customComponents":{}}
  </div>
  
  
</div>
`;

exports[`svelte/Field renders correctly - DynamicObject 2`] = `
<div>
  <div
    id="field"
  >
    {"id":"path__0__to__field","path":"path.0.to.field","fields":[],"minItems":0,"modifiers":"success ","isActive":true,"maxItems":null,"variables":{"var":"value"},"userInputs":{"test":"ok"},"label":null,"helper":null,"allowedPatterns":[],"addButtonProps":{"label":null},"removeButtonProps":{"label":null},"addTextfieldProps":{"label":"TRANSLATED LABEL","placeholder":"TRANSLATED LABEL"},"type":"dynamicObject","value":null,"customComponents":{}}
  </div>
  
  
</div>
`;

exports[`svelte/Field renders correctly - FilePicker 1`] = `
<div>
  <div>
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    path.0.to.field
     
    null
     
    undefined
     
    null
     
    null
     
    undefined
     
    path__0__to__field
     
    false
     
    success 
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
  </div>
  
  
</div>
`;

exports[`svelte/Field renders correctly - FilePicker 2`] = `
<div>
  <div>
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    path.0.to.field
     
    null
     
    undefined
     
    TRANSLATED LABEL
     
    null
     
    undefined
     
    path__0__to__field
     
    false
     
    success 
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
  </div>
  
  
</div>
`;

exports[`svelte/Field renders correctly - Link 1`] = `
<div>
  <div>
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    path__0__to__field
     
    undefined
     
    success 
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
  </div>
  
  
</div>
`;

exports[`svelte/Field renders correctly - Message with label and helper 1`] = `
<div>
  <div
    id="field"
  >
    {"id":"path__0__to__field","label":"TRANSLATED LABEL","modifiers":"success "}
  </div>
  
  
</div>
`;

exports[`svelte/Field renders correctly - Message with label and message 1`] = `
<div>
  <div
    id="field"
  >
    {"id":"path__0__to__field","label":"TRANSLATED LABEL","modifiers":"success "}
  </div>
  
  
</div>
`;

exports[`svelte/Field renders correctly - Options 1`] = `
<div>
  <div>
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    path.0.to.field
     
    null
     
    undefined
     
    undefined
     
    null
     
    undefined
     
    path__0__to__field
     
    undefined
     
    success 
     
    undefined
     
    undefined
     
    () =&gt; null
     
    [object Object],[object Object]
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
  </div>
  
  
</div>
`;

exports[`svelte/Field renders correctly - Textarea 1`] = `
<div>
  <div>
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    path.0.to.field
     
    null
     
    null
     
    null
     
    null
     
    100
     
    path__0__to__field
     
    false
     
    success 
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
  </div>
  
  
</div>
`;

exports[`svelte/Field renders correctly - Textarea 2`] = `
<div>
  <div>
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    path.0.to.field
     
    null
     
    test
     
    TRANSLATED LABEL
     
    null
     
    10
     
    path__0__to__field
     
    false
     
    success 
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
  </div>
  
  
</div>
`;

exports[`svelte/Field renders correctly - Textfield 1`] = `
<div>
  <div>
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    path.0.to.field
     
    null
     
    null
     
    null
     
    null
     
    100
     
    path__0__to__field
     
    false
     
    success 
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
  </div>
  
  
</div>
`;

exports[`svelte/Field renders correctly - Textfield 2`] = `
<div>
  <div>
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    path.0.to.field
     
    null
     
    test
     
    TRANSLATED LABEL
     
    null
     
    10
     
    path__0__to__field
     
    false
     
    success 
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
     
    undefined
  </div>
  
  
</div>
`;

exports[`svelte/Field renders correctly - Unknown component 1`] = `
<div>
  
</div>
`;
